File: AFMobile/AFWinPhone/AFWinPhone/builders/widgets/DateWidgetBuilder.cs
﻿using AFWinPhone.components.parts;
using AFWinPhone.builders.skins;
using AFWinPhone.utils;
using System;
using System.Globalization;
using Windows.UI;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Media;

namespace AFWinPhone.builders.widgets
{
    class DateWidgetBuilder : BasicBuilder
    {
        private String dateFormat;

        public DateWidgetBuilder(Skin skin, FieldInfo properties) : base(skin, properties)
        {
            this.dateFormat = "dd.MM.yyyy"; //Default date format
        }


        public override FrameworkElement buildFieldView()
        {
            StackPanel dateLayout = new StackPanel();
            dateLayout.Orientation = Orientation.Horizontal;

            DatePicker datePicker = new DatePicker();
            datePicker.VerticalAlignment = VerticalAlignment.Top;
            datePicker.HorizontalAlignment = HorizontalAlignment.Stretch;

            datePicker.FontFamily = getSkin().getFieldFont();
            datePicker.FontSize = getSkin().getFieldFontSize();
            //note .. do not set color - should be according to wp theme

            if (getProperties().isReadOnly())
            {
                datePicker.IsEnabled = false;
                datePicker.Foreground = new SolidColorBrush(Colors.LightGray);
            }
            return datePicker;
        }

        public override void setData(AFField field, Object value)
        {
            DatePicker dateText = (DatePicker)field.getFieldView();
            DateTime? date = Utils.ParseDate(value.ToString());
            if (date != null)
            {
                dateText.Date = date.Value;
                field.setActualData(date.Value.ToString("dd.MM.yyyy", CultureInfo.InvariantCulture));
            }
            else {
                //parsing totally failed maybe exception
            }
        }

        public override Object getData(AFField field)
        {
            DatePicker dateText = (DatePicker)field.getFieldView();
            DateTimeOffset? date = dateText.Date;
            if (date != null)
            {
                return date.Value.ToString("yyyy-MM-dd'T'HH:mm:ss", CultureInfo.InvariantCulture);
            }
            return null;
        }
    }
}

File: AFMobile/AFAndroid/app/src/main/java/cz/cvut/fel/matyapav/afandroid/builders/widgets/DateWidgetBuilder.java
package cz.cvut.fel.matyapav.afandroid.builders.widgets;

import android.app.DatePickerDialog;
import android.content.Context;
import android.graphics.Color;
import android.text.InputType;
import android.view.View;
import android.view.ViewGroup;
import android.widget.DatePicker;
import android.widget.EditText;
import android.widget.LinearLayout;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;

import cz.cvut.fel.matyapav.afandroid.builders.skins.Skin;
import cz.cvut.fel.matyapav.afandroid.components.parts.AFField;
import cz.cvut.fel.matyapav.afandroid.components.parts.FieldInfo;
import cz.cvut.fel.matyapav.afandroid.utils.Utils;

/**
 * @author Pavel Matyáš (matyapav@fel.cvut.cz).
 *
 *@since 1.0.0..
 */
public class DateWidgetBuilder extends BasicBuilder {

    private String dateFormat;

    DateWidgetBuilder(Context context, Skin skin, FieldInfo properties) {
        super(context, skin, properties);
        this.dateFormat = "dd.MM.yyyy"; //Default date format
    }

    @Override
    public View buildFieldView(final Context context) {
        LinearLayout dateLayout = new LinearLayout(context);
        dateLayout.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT));
        dateLayout.setOrientation(LinearLayout.HORIZONTAL);
        final EditText dateText = new EditText(context);
        dateText.setLayoutParams(new LinearLayout.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT));
        dateText.setTextColor(getSkin().getFieldColor());
        dateText.setTypeface(getSkin().getFieldFont());
        dateText.setHint(dateFormat);
        dateText.setFocusable(false);
        dateText.setClickable(true);

        //TODO umoznit clear
        dateText.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final SimpleDateFormat dateFormatter = new SimpleDateFormat(dateFormat, Locale.getDefault());
                Calendar newCalendar = Calendar.getInstance();
                DatePickerDialog fromDatePickerDialog = new DatePickerDialog(context, new DatePickerDialog.OnDateSetListener() {
                    @Override
                    public void onDateSet(DatePicker view, int year, int monthOfYear, int dayOfMonth) {
                        Calendar newDate = Calendar.getInstance();
                        newDate.set(year, monthOfYear, dayOfMonth);
                        dateText.setText(dateFormatter.format(newDate.getTime()));
                    }
                }, newCalendar.get(Calendar.YEAR), newCalendar.get(Calendar.MONTH), newCalendar.get(Calendar.DAY_OF_MONTH));
                fromDatePickerDialog.show();
            }
        });

        if(getProperties().isReadOnly()){
            dateText.setInputType(InputType.TYPE_NULL);
            dateText.setTextColor(Color.LTGRAY);
        }
        return dateText;
    }

    @Override
    public void setData(AFField field, Object value) {
        EditText dateText = (EditText) field.getFieldView();
        SimpleDateFormat outputFormatter = new SimpleDateFormat("dd.MM.yyyy", Locale.getDefault());
        Date date = Utils.parseDate(String.valueOf(value));
        if(date != null) {
            dateText.setText(outputFormatter.format(date));
            field.setActualData(outputFormatter.format(date));
        } else {
            System.err.println("Date " + value + " parsing failed.");
        }
    }

    @Override
    public Object getData(AFField field) {
        EditText dateText = (EditText) field.getFieldView();
        SimpleDateFormat serverFormatter = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSZ", Locale.getDefault());
        Date date = Utils.parseDate(dateText.getText().toString());
        if(date != null) {
            return serverFormatter.format(date);
        }
        return null;
    }


}
